x-pack/plugins/security_solution/public/detection_engine/rule_creation/components/esql_query_edit/validators/esql_query_validator_factory.test.ts

```diff
@@ -0,0 +1,182 @@
+      ['from test* | keep metadata'],
+      ['from test* | eval x="metadata _id"'],
+    ])('reports when METADATA operator is missing in a NON aggregating query "%s"', (esqlQuery) =>
+      expect(
+        createValidator()({
+          value: createEsqlQueryFieldValue(esqlQuery),
+        } as EsqlQueryValidatorArgs)
+      ).resolves.toMatchObject({
+        code: ESQL_ERROR_CODES.ERR_MISSING_ID_FIELD_FROM_RESULT,
+      })
+    );
+
+    it.each([
+      ['from test* metadata _id'],
+      ['from test* metadata _id, _index'],
+      ['from test* metadata _index, _id'],
+      ['from test*  metadata _id '],
+      ['from test*    metadata _id '],
+      ['from test*  metadata _id | limit 10'],
+      [
+        `from packetbeat* metadata
+
+        _id
+        | limit 100`,
+      ],
+    ])(
+      'succeeds validation when METADATA operator EXISTS in a NON aggregating query "%s"',
+      (esqlQuery) =>
+        expect(
+          createValidator()({
+            value: createEsqlQueryFieldValue(esqlQuery),
+          } as EsqlQueryValidatorArgs)
+        ).resolves.toBeUndefined()
+    );
+
+    it('succeeds validation when METADATA operator is missing in an aggregating query "from test* | stats c = count(*) by fieldA"', () =>
+      expect(
+        createValidator()({
+          value: createEsqlQueryFieldValue('from test* | stats c = count(*) by fieldA'),
+        } as EsqlQueryValidatorArgs)
+      ).resolves.toBeUndefined());
+  });
+
+  describe('METADATA _id field validation for NON aggregating queries', () => {
+    it('reports when METADATA "_id" field is missing', () => {
+      getESQLQueryColumnsMock.mockResolvedValue([{ id: 'column1' }, { id: 'column2' }]);
+
+      return expect(
+        createValidator()({
+          value: createEsqlQueryFieldValue('from test*'),
+        } as EsqlQueryValidatorArgs)
+      ).resolves.toMatchObject({
+        code: ESQL_ERROR_CODES.ERR_MISSING_ID_FIELD_FROM_RESULT,
+      });
+    });
+
+    it('succeeds validation when METADATA "_id" field EXISTS', async () => {
+      getESQLQueryColumnsMock.mockResolvedValue([{ id: '_id' }, { id: 'column1' }]);
+
+      return expect(
+        createValidator()({
+          value: createEsqlQueryFieldValue('from test* metadata _id'),
+        } as EsqlQueryValidatorArgs)
+      ).resolves.toBeUndefined();
+    });
+  });
+
+  describe('METADATA _id field validation for aggregating queries', () => {
+    it('succeeds validation when METADATA operator with "_id" field is missing', () => {
+      getESQLQueryColumnsMock.mockResolvedValue([{ id: 'column1' }, { id: 'column2' }]);
+
+      return expect(
+        createValidator()({
+          value: createEsqlQueryFieldValue(
+            'from test* metadata someField | stats c = count(*) by fieldA'
+          ),
+        } as EsqlQueryValidatorArgs)
+      ).resolves.toBeUndefined();
+    });
+  });
+
+  describe('when getESQLQueryColumns fails', () => {
+    it('returns an empty array', () => {
```

Comment:
I don't see any checks for an empty array, it this description relevant?

Reply:
fixed